### PR TITLE
DynamicGroups filter data migration from Site/Region to Location

### PR DESCRIPTION
**Pull request by HanlinMiao**
# Closes: #<ISSUE NUMBER GOES HERE>
# What's Changed

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example Plugin Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/dcim/migrations/0030_migrate_region_and_site_data_to_locations.py [hunk 1]
```diff
@@ -239,6 +239,17 @@ def migrate_site_and_region_data_to_locations(apps, schema_editor):
...
 
+        DynamicGroup = apps.get_model("extras", "DynamicGroup")
+        dynamic_groups = DynamicGroup.objects.all()
+        for dg in dynamic_groups:
+            if "location" not in dg.filter.keys():
+                dg.filter["location"] = []
+            dg.filter["location"] += dg.filter.get("region", [])
+            dg.filter.pop("region", None)
+            if not dg.filter["location"]:  # Remove the location key if the list is empty
+                dg.filter.pop("location")
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0030_migrate_region_and_site_data_to_locations.py:250]: Maybe?

```python
if "region" in dg.filter:
    dg.filter.setdefault("location", []).extend(dg.filter.pop("region"))
```

**Review by glennmatthews — APPROVED** [review]
Looks valid to me.

This makes me think that it would be a good feature for us to introduce something like a `nautobot-server audit_dynamic_groups` command which would iterate over all DGs and make sure their filters are valid. Because we're not currently fixing up DGs to account for all of the other filtering changes that we're making in 2.0...

File: nautobot/dcim/migrations/0030_migrate_region_and_site_data_to_locations.py [hunk 2]
```diff
@@ -239,6 +239,15 @@ def migrate_site_and_region_data_to_locations(apps, schema_editor):
 
         # Custom Links and Image Attachements do not have Region ContentType as one of its ContentType options
 
+        DynamicGroup = apps.get_model("extras", "DynamicGroup")
+        dynamic_groups = DynamicGroup.objects.all()
+        for dg in dynamic_groups:
+            if "region" in dg.filter:
+                dg.filter.setdefault("location", []).extend(dg.filter.pop("region"))
+            if not dg.filter["location"]:  # Remove the location key if the list is empty
+                dg.filter.pop("location")
```
Comment (glennmatthews) [inline nautobot/dcim/migrations/0030_migrate_region_and_site_data_to_locations.py:248]: Should these two lines be removed? Same question below.

Comment (HanlinMiao) [reply nautobot/dcim/migrations/0030_migrate_region_and_site_data_to_locations.py:248]: Right right they should, I was thinking that "location" key might be an empty list, but we are already checking the "region" key to ensure the list wont be empty. Sorry!

**Pull request merged**